src/object/cli_obj.c

```diff
@@ -1530,49 +1470,496 @@ obj_shard_task_sched(struct obj_auxi_args *obj_auxi)
 		tse_task_complete(obj_auxi->obj_task, 0);
 }
 
-struct shard_reset_target {
-	unsigned int		 srt_is_punch:1;
-	unsigned int		 srt_rdg_count;
-	unsigned int		*srt_shards;
-};
+static struct shard_auxi_args *
+obj_embedded_shard_arg(struct obj_auxi_args *obj_auxi)
+{
+	switch (obj_auxi->opc) {
+	case DAOS_OBJ_RPC_UPDATE:
+	case DAOS_OBJ_RPC_FETCH:
+		return &obj_auxi->rw_args.auxi;
+	case DAOS_OBJ_RPC_PUNCH:
+	case DAOS_OBJ_RPC_PUNCH_DKEYS:
+	case DAOS_OBJ_RPC_PUNCH_AKEYS:
+		return &obj_auxi->p_args.pa_auxi;
+	default:
+		D_ERROR("bad opc %d.\n", obj_auxi->opc);
+		return NULL;
+	};
+}
+
+static int
+shard_io(tse_task_t *task, struct shard_auxi_args *shard_auxi)
+{
+	struct dc_object		*obj = shard_auxi->obj;
+	struct obj_auxi_args		*obj_auxi = shard_auxi->obj_auxi;
+	struct dc_obj_shard		*obj_shard;
+	struct obj_req_tgts		*req_tgts;
+	struct daos_obj_shard_tgt	*fw_shard_tgts;
+	uint32_t			 fw_cnt;
+	int				 rc;
+
+	D_ASSERT(obj != NULL);
+	rc = obj_shard_open(obj, shard_auxi->shard, shard_auxi->map_ver,
+			    &obj_shard);
+	if (rc != 0) {
+		/* skip a failed target */
+		if (rc == -DER_NONEXIST)
+			rc = 0;
+
+		/* complete it if it is shard task */
+		if (!d_list_empty(&obj_auxi->shard_task_head))
```

Comment:
hmm, even for single shard case, it can still complete the task? no?

Reply:
ah, i see. previously I complete the task after calling it, later I change that but forgot to remove this here. I'll fix it soon, thanks.